[Gradient Compression] Allow BatchedPowerSGD to run vanilla allreduce for the first K iterations

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51427 [Gradient Compression] Check start_PowerSGD_iter > 1
* **#51270 [Gradient Compression] Allow BatchedPowerSGD to run vanilla allreduce for the first K iterations**
* #51400 [Resubmission][Gradient Compression] Refactor default_hooks.py and powerSGD_hook.py by creating a util function that make a vanilla allreduce future

Similar to #50973, allow the batched version to run vanilla allreduce for the first K iterations.

This may be useful if the batched version can be applied to some use cases where the accuracy requirement is not very strict.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D26077709](https://our.internmc.facebook.com/intern/diff/D26077709/)

Commit:
[Gradient Compression] Allow BatchedPowerSGD to run vanilla allreduce for the first K iterations

Similar to #50973, allow the batched version to run vanilla allreduce for the first K iterations.

This may be useful if the batched version can be applied to some use cases where the accuracy requirement is not very strict.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D26077709](https://our.internmc.facebook.com/intern/diff/D26077709/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -423,6 +423,13 @@ def batched_powerSGD_hook(state: PowerSGDState, bucket) -> torch.futures.Future:
     input_tensor = bucket.get_tensors()[0]
+
+    # Run vanilla allreduce in the first `start_powerSGD_iter` iterations.
+    if state.iter < state.start_powerSGD_iter:
```
Do we have unittests in place for `batched_powerSGD_hook`? If not, can we add them?

SciPioneer replied:
Batched powerSGD is already tested by `test_powerSGD_ddp_comm_hook_nccl` in test_c10d.py.

Commit:
Update on "[Gradient Compression] Allow BatchedPowerSGD to run vanilla allreduce for the first K iterations"


Similar to #50973, allow the batched version to run vanilla allreduce for the first K iterations.

This may be useful if the batched version can be applied to some use cases where the accuracy requirement is not very strict.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D26077709](https://our.internmc.facebook.com/intern/diff/D26077709/)

[ghstack-poisoned]

rohan-varma reviewed (commented):
I think the diff below this one was reverted, and this one is also failing tests: https://app.circleci.com/pipelines/github/pytorch/pytorch/266089/workflows/a036f791-01c8-4538-90eb-a4e40234b8c3/jobs/10492151. Do you need to resubmit the previous diff first?

SciPioneer commented:
> I think the diff below this one was reverted, and this one is also failing tests: https://app.circleci.com/pipelines/github/pytorch/pytorch/266089/workflows/a036f791-01c8-4538-90eb-a4e40234b8c3/jobs/10492151. Do you need to resubmit the previous diff first?

Created #51400 for the resubmission, and will submit that PR first.

Commit:
Update on "[Gradient Compression] Allow BatchedPowerSGD to run vanilla allreduce for the first K iterations"


Similar to #50973, allow the batched version to run vanilla allreduce for the first K iterations.

This may be useful if the batched version can be applied to some use cases where the accuracy requirement is not very strict.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D26077709](https://our.internmc.facebook.com/intern/diff/D26077709/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Allow BatchedPowerSGD to run vanilla allreduce for the first K iterations"


Similar to #50973, allow the batched version to run vanilla allreduce for the first K iterations.

This may be useful if the batched version can be applied to some use cases where the accuracy requirement is not very strict.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D26077709](https://our.internmc.facebook.com/intern/diff/D26077709/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -399,6 +399,10 @@ def batched_powerSGD_hook(state: PowerSGDState, bucket) -> torch.futures.Future:
     7) Computes M, which is approximately equal to PQ^T.
     8) Truncates the input tensor to the original length.
 
+    Note that this communication hook enforces vanilla allreduce for the first `state.start_powerSGD_iter` iterations.
```
Do we have the default value for `state.start_powersgd_iter` documented in the docs for powerSGDState? Would be nice to ensure we have that, and maybe also specify the default start iteration here.
